proposals/policy_status.md

```diff
@@ -0,0 +1,183 @@
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+  - [Ready Column Modification](#ready-column-modification)
+  - [Dependency Matrix](#dependency-matrix)
+- [Implementation](#implementation)
+- [Migration](#migration-optional)
+- [Alternatives](#alternatives)
+- [CRD Changes](#crd-changes-optional)
+
+# Overview
+
+[overview]: #overview
+
+This proposal introduces a new system for Kyverno to report the operational readiness of its policies. The system provides detailed status conditions for each policy, indicating whether webhooks are configured correctly (for Admission controllers), if the policy is loaded in Kyverno's cache, and if Kyverno has the necessary RBAC permissions to enforce it. Additionally, it proposes a modification to the `kubectl` output to display the number of ready conditions over the total number of relevant conditions. This will enhance troubleshooting, improve reliability, and provide better control over Kyverno's operation.
+
+# Definitions
+
+[definitions]: #definitions
+
+- **Policy:** A set of rules in Kyverno that define desired configurations or behaviors for Kubernetes resources.
+- **Webhook:** A mechanism in Kubernetes that allows external services (like Kyverno) to intercept and potentially modify API requests.
+- **Admission Controller:** A type of Kyverno controller that intercepts and processes API requests before they are persisted to the Kubernetes cluster.
+- **Background Controller:** A type of Kyverno controller that processes resources after they are created or updated.
+- **Cache:** An internal data store used by Kyverno to quickly access frequently used information, such as policies.
+- **RBAC (Role-Based Access Control):** A Kubernetes system for controlling who can access and modify resources.
+- **CRD (Custom Resource Definition):** A way to extend Kubernetes with custom resource types.
+- **NotReady:** A condition type that indicates that the specific sub-system is not ready and needs attention.
+
+# Motivation
+
+[motivation]: #motivation
+
+- **Why should we do this?** Currently, diagnosing issues with Kyverno policies can be challenging due to limited status information. This lack of clarity leads to wasted time and effort during troubleshooting.
+- **What use cases does it support?**
+  - Quickly identifying the root cause of policy failures (e.g., webhook misconfiguration, missing permissions).
+  - Automating monitoring and alerting based on policy readiness.
+  - Improving the overall reliability of policy enforcement.
+  - Gaining a better understanding of Kyverno's operational state.
+- **What is the expected outcome?** Clear, actionable status information for each policy will enable users to quickly resolve issues, prevent policy deployment failures, and ensure that policies are enforced as intended. The modification to the `kubectl` output will provide a quick overview of policy readiness at a glance.
+
+# Proposal
+
+[proposal]: #proposal
```

Comment:
I recommend we keep the `status` field with values `Ready` and `NotReady` as a summary state. We can set `status: Ready` when all conditions pass.

Reply:
So, first I think it's deprecated. I will try to get a reference to this. Second, for future proofing I was thinking that different components might interpret conditions differently. So for instance, the background controller might not care about a policy being configured in the webhook or not. Ready here should be false for Admission but true for Background.

I think it's better if we leave each component to interpret the readiness on its own. But, I'm also fine with this.